### PR TITLE
Lower contention LruCache

### PR DESCRIPTION
**Pull request by benaadams**
## Changes

- Use ReadWriteLock so many readers can access the LruCache at the same time
- Queue the LinkedList and overflow changes to a single background ThreadPoolItem

#### What types of changes does your code introduce?

- [x] Optimization

#### Requires testing

- [ ] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by LukaszRozmej** [comment]
Not sure about the design, wouldn't concurrent dictionary be better in terms of Set's?
Also should we replace the current LRU, or just create another one and benchmark them in different situations?

File: src/Nethermind/Nethermind.Core/Caching/ConcurrentLruCache.cs [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+            _leastRecentlyUsed = null;
+        }
+
+        public TValue Get(TKey key) => TryGet(key, out TValue value) ? value : default!;
+
+        public bool TryGet(TKey key, out TValue value)
+        {
+            bool didGet;
+            LinkedListNode<LruCacheItem>? node;
+            using (var handle = _lock.AcquireWrite())
```
Comment (Scooletz) [inline src/Nethermind/Nethermind.Core/Caching/ConcurrentLruCache.cs:69]: `AcquireWrite`? Is it intentional?

File: src/Nethermind/Nethermind.Core/Caching/ConcurrentLruCache.cs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        {
+            if (val is null)
+            {
+                Delete(key);
+            }
+
+            ref LinkedListNode<LruCacheItem>? node = ref Unsafe.NullRef<LinkedListNode<LruCacheItem>?>();
+            using (var handle = _lock.AcquireWrite())
+            {
+                node = ref CollectionsMarshal.GetValueRefOrAddDefault(_cacheMap, key, out _);
```
Comment (Scooletz) [inline src/Nethermind/Nethermind.Core/Caching/ConcurrentLruCache.cs:95]: This is a ref to an array. If another thread resizes the array, this ref will point to an array no longer in use. Am I missing something or is it a bug?

File: src/Nethermind/Nethermind.Core/Caching/ConcurrentLruCache.cs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        }
+
+        private void ProcessExpiredItems()
+        {
+            CancellationToken ct = _cts.Token;
+            while (true)
+            {
+                while (!ct.IsCancellationRequested && _accesses.TryDequeue(out LinkedListNode<LruCacheItem>? node))
+                {
+                    bool exists = _cacheMap.ContainsKey(node.Value.Key);
```
Comment (Scooletz) [inline src/Nethermind/Nethermind.Core/Caching/ConcurrentLruCache.cs:178]: Can be broken as `_cacheMap` might be written to concurrently?

File: src/Nethermind/Nethermind.Core/Caching/NonBlockingLruCache.cs [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+        public bool SetResult(TKey key, TValue val)
+        {
+            if (val is null)
+            {
+                return Delete(key);
+            }
+
+            bool exists = false; // Captured by lambda
+            LinkedListNode<LruCacheItem> node = _cacheMap.AddOrUpdate(
```
Comment (Scooletz) [inline src/Nethermind/Nethermind.Core/Caching/NonBlockingLruCache.cs:105]: Do we always allocate an item on non-existence of the key here and not reuse the recent as thsi PR separates this two? Just asking, can be by design.

File: src/Nethermind/Nethermind.Core/Caching/NonBlockingLruCache.cs [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+                while (!ct.IsCancellationRequested && _cacheMap.Count > _maxCapacity)
+                {
+                    LinkedListNode<LruCacheItem>? leastRecentlyUsed = _leastRecentlyUsed;
+                    if (leastRecentlyUsed is null)
+                    {
+                        // Clear has been called
+                        break;
+                    }
+
+                    _cacheMap.Remove(leastRecentlyUsed.Value.Key, out leastRecentlyUsed);
```
Comment (Scooletz) [inline src/Nethermind/Nethermind.Core/Caching/NonBlockingLruCache.cs:210]: I see that `.Remove` is called in two places only, one on `Delete` and second in here.

File: src/Nethermind/Nethermind.Core/Caching/NonBlockingLruCache.cs [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Nethermind.Core.Extensions;
+
+namespace Nethermind.Core.Caching
+{
+    using NonBlocking;
+
+    public sealed class NonBlockingLruCache<TKey, TValue> : IThreadPoolWorkItem, ICache<TKey, TValue> where TKey : notnull
```
Comment (Scooletz) [inline src/Nethermind/Nethermind.Core/Caching/NonBlockingLruCache.cs:15]: I see how you split implementations now. Why not to hide them behind `LruCache` generic class and return one depending on the `Unsafe.SizeOf<TKey>` ?

**Review by Scooletz — COMMENTED** [review]
I read through the PR and provided some comments. 

As a general one, this PR makes the cleaning separate and results in potentially occasionally bigger dictionaries. This separation makes me wonder, whether we could drop the linkage of `LinkedListNode` altogether and 

1. create `Node<T>` that would have a `T Value` and `long Epoch`. 
2. on operation, you `var epoch = Interlocked.Increment(ref _counter)` and set the `Epoch` in the item
3. you enqueue `(node, epoch)`
4. during cleanup, you pull enqueue items, and check `tuple.node.Epoch == tuple.epoch`, if right and the map is to big you remove the thing. Otherwise you discard because something else touched this.

No allocations for LinkedListItem at all, as `_accesses` store now value tuples, in a similar way as `_cacheMap`.

This would make is similar to `ConcurrentQueue` a bit 😅 using the epoch thing.

WDYT @benaadams ?

**Pull request closed without merging**